### PR TITLE
External operators

### PR DESCRIPTION
**Pull request by nbouziani**
This PR contains the Firedrake implementation of external operators. The corresponding UFL implementation was merged in https://github.com/FEniCS/ufl/pull/206. The Firedrake infrastructure to make external operators has already been introduced in #2297, making that PR pretty "small".

The external operator interface allows for incorporating arbitrary operators mapping between finite element spaces in Firedrake. Those operators are external in the sense that their implementation is left to be specified by the user. This PR mainly contains:

- The assembly mechanism required for evaluating external operators, and expressions/variational forms containing external operators.
- The abstract external operator class that orchestrates the assembly
- The ML operator which allows for incorporating machine learning models implemented in PyTorch in Firedrake
- A new manual chapter

### CODE REVIEW

**Comment by nbouziani** [comment]
Docstrings and manual to be updated/completed soon.

File: firedrake/external_operators/ml_operator.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+         """
+        return self.operator_data['inputs_format']
+
+    def _ufl_expr_reconstruct_(self, *operands, function_space=None, derivatives=None, argument_slots=(), operator_data=None):
+        "Overwrite _ufl_expr_reconstruct to pass on params_version"
+        return AbstractExternalOperator._ufl_expr_reconstruct_(self, *operands, function_space=function_space,
+                                                               derivatives=derivatives,
+                                                               argument_slots=argument_slots,
+                                                               operator_data=operator_data)
+
```
Comment (nbouziani) [inline firedrake/external_operators/ml_operator.py:47]:
```suggestion
```

File: firedrake/external_operators/abstract_external_operators.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+import firedrake.ufl_expr as ufl_expr
+from firedrake.assemble import allocate_matrix
+from firedrake.function import Function
+from firedrake.cofunction import Cofunction
+from firedrake.matrix import MatrixBase
+from firedrake import functionspaceimpl
+
+
+class RegisteringAssemblyMethods(UFLType):
```
Comment (dham) [inline firedrake/external_operators/abstract_external_operators.py:13]: AssemblyRegisterMetaClass

Comment (dham) [reply firedrake/external_operators/abstract_external_operators.py:13]: Please document this dark magic.

File: firedrake/external_operators/abstract_external_operators.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+        action(adjoint(dNdu))                  | dN/du^{*}(u, m; vv, uhat)        | (1, 0) | (1, None)
+        action(adjoint(dNdm))                  | dN/dm^{*}(u, m; vv, mhat)        | (0, 1) | (1, None)
+                                               |                                  |        |
+        d2Ndu = derivative(dNdu, u, uhat)      |                                  |        |
+                                               |                                  |        |
+        action(d2Ndu, uu)                      | d2N/dudu(u, m; uu, uhat, v*)     | (2, 0) | (0, 1, None)
+        adjoint(action(d2Ndu, uu))             | d2N/dudu^{*}(u, m; v*, uhat, uu) | (2, 0) | (None, 1, 0)
+        action(adjoint(action(d2Ndu, uu)), vv) | d2N/dudu^{*}(u, m; vv, uhat, uu) | (2, 0) | (None, 1, None)
+
+        Here are examples on how to specify the implementation of:
```
Comment (dham) [inline firedrake/external_operators/abstract_external_operators.py:101]: Lift to manual

File: firedrake/external_operators/ml_operator.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    # -- Assembly methods -- #
+
+    @assemble_method(0, (0,))
+    def assemble_model(self, *args, **kwargs):
+        return self._forward()
+
+    @assemble_method(1, (0, 1))
+    def assemble_jacobian(self, *args, assembly_opts, **kwargs):
+        # Get Jacobian using PyTorch AD
+        J = self._jac()
```
Comment (dham) [inline firedrake/external_operators/ml_operator.py:50]: pytorch?

File: firedrake/external_operators/point_expr_operator.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    def __init__(self, *operands, function_space, derivatives=None, argument_slots=(), operator_data):
+        r"""
+        :param operands: operands on which act the :class:`PointexrOperator`.
+        :param function_space: the :class:`.FunctionSpace`,
+        or :class:`.MixedFunctionSpace` on which to build this :class:`Function`.
+        Alternatively, another :class:`Function` may be passed here and its function space
+        will be used to build this :class:`Function`.  In this case, the function values are copied.
+        :param derivatives: tuple specifiying the derivative multiindex.
+        :param operator_data: dictionary containing the function defining how to evaluate the :class:`PointexprOperator`.
+        """
```
Comment (dham) [inline firedrake/external_operators/point_expr_operator.py:27]: Numpydoc

File: firedrake/ml/pytorch/neural_network_operators.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    # using `torch.autograd.functional.{hvp, hessian, vhp}`
+
+    # `vjp` is faster than `.backward` and give you model output + vector-Jacbian product at same time in 1 traversal
+
+    # -- PyTorch routines for computing AD based quantities via `torch.autograd.functional` -- #
+
+    def _vjp(self, δy):
+        model = self.model
+        x = self._pre_forward_callback(*self.ufl_operands)
+        δy_P = self._pre_forward_callback(δy)
```
Comment (dham) [inline firedrake/ml/pytorch/neural_network_operators.py:103]: Don't use greek!!!!!!!

**Review by dham — CHANGES_REQUESTED** [review]

**Review by dham — APPROVED** [review]

**Pull request merged**